pytorch_lightning/utilities/apply_func.py

```diff
@@ -110,6 +111,14 @@ def apply_to_collection(
                 out.append(v)
         return elem_type(*out) if is_namedtuple else elem_type(out)
 
+    if dataclasses.is_dataclass(data) and not isinstance(data, type):
```

Comment:
why the `and not isinstance(data, type)` check?

Reply:
This is the official recommended way:
> https://docs.python.org/3/library/dataclasses.html#module-level-decorators-classes-and-functions
> 
> dataclasses.is_dataclass(class_or_instance)
> Return True if its parameter is a dataclass or an instance of one, otherwise return False.
> If you need to know if a class is an instance of a dataclass (and not a dataclass itself), then add a further check for not isinstance(obj, type):
> ```python
> def is_dataclass_instance(obj):
>     return is_dataclass(obj) and not isinstance(obj, type)
> ```

Reply:
I think it would be ncie for you to copy that function and use it, just as we have for `_is_namedtuple`